### PR TITLE
#781 - accounts linking

### LINKED ISSUE #781: Auth0 user account linking (v1.1+)
We don't allow duplicate emails so if someone tries to signup with a social account and the same email already registered under they get an error (see #772). 

Eventually we want to enable user account linking in such a case. Not critical - release v1.1 (or later). 

https://auth0.com/docs/users/concepts/overview-user-account-linking

### LINKED ISSUE #1266: User able to signup again with same emailid by using facebook,linkedin,google.
Environment: https://rcv11.fightpandemics.xyz/
Device used: Samsung galaxy s8
Browser: Google chrome
Steps to create:
1.Signup with a mail id(ex:abc@gmail.com).
2.The user have a facebook account with the same email id abc@gmail.com.
3.If the user tries to signup with facebook account again.
Expected Result:
It should not allow user to create new account with same email id.
Actual Result:
It allows user to create different accounts with facebook and manual creation with same email id.
But it doesnt allow to create a profile.(It idetifies the match in emailid only at profile creation page)
Screenshot_20200723-164037_Chrome

### PR DESCRIPTION
**Pull request by ddveloper**
### What does this pull request aim to achieve? 💯

_Please be concise and link any related issue(s):_

  to solve #781 account linking

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix, where `<branch_name>` briefly describes the issue(s) resolved and is prefixed with the issue number(s) (e.g. `1127-1130-update-git-branching-model`). The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".
- [x] The title describes the issue(s) being addressed using format: **`<issue_numbers> - <brief_issue_description>`** (e.g. `#1127/#1130 - Update Git Branching Model`)

### CODE REVIEW

File: backend/lib/components/Auth0.js
Comment (joshmorel) [inline backend/lib/components/Auth0.js]:
```suggestion
const _updateUser = async (token, userId, payload) => {
```

Not sure what you're planning to use this for but we actually are just about to implement something very similar here:

https://github.com/FightPandemics/FightPandemics/blob/86883fe123b01991c6488fa966e9c68a331267db/backend/lib/components/Auth0.js#L88

It is very close to getting ready to merge so maybe we should wait on that first?

File: backend/lib/components/Auth0.js
Comment (joshmorel) [inline backend/lib/components/Auth0.js]:
```suggestion
```

I don't believe we are planning to add functionality to unlink account so I think we shouldn't include this until that point.

In some rare case where we need to we can do so manually in Auth0 on request.

**Review by joshmorel — CHANGES_REQUESTED** [review]
A few changes to request for now I will do full review later (more question in Slack).

File: backend/lib/components/Auth0.js
Comment (joshmorel) [inline backend/lib/components/Auth0.js]: I don't think we'll need to worry about merging metadata.

From https://auth0.com/docs/users/user-account-linking:

> The user_metadata and app_metadata of the secondary account are discarded

The secondary account in our flow should have no metadata so there's nothing to worry about.

**Review by joshmorel — CHANGES_REQUESTED** [review]

File: backend/lib/components/Auth0.js
Comment (ddveloper) [reply backend/lib/components/Auth0.js]: I don't know if our QA team will need such a function to `unlink` or even `delete` accounts, to facilitate the automation test. and I agree it's not what we need for this feature, I will remove it

File: backend/lib/components/Auth0.js
Comment (ddveloper) [reply backend/lib/components/Auth0.js]: You are right Josh, I don't know if we are going to use metadata later or not, but for this feature, we don't need it, I will remove it (keep it somewhere in case we need it later : )

File: backend/lib/components/Auth0.js
Comment (ddveloper) [reply backend/lib/components/Auth0.js]: this is for merging the meta-data only, since we're removing that part, I will also remove this one

**Comment by ddveloper** [comment]
the review build URL: https://feature781-linkaccounts.fightpandemics.xyz/

File: backend/lib/components/Auth0.js
Comment (mannykary) [inline backend/lib/components/Auth0.js]: What is lodash being used for? I don't see it being used anywhere in the code you just added. If it is not being used, then we should remove this import.

File: backend/lib/endpoints/auth.js
Comment (mannykary) [inline backend/lib/endpoints/auth.js]: Probably could just use dot notation: `payload.sub`

File: backend/lib/endpoints/auth.js
Comment (mannykary) [inline backend/lib/endpoints/auth.js]: Similar to previous comment, could probably use dot notation

File: client/src/pages/CreateUserProfile.js
Comment (mannykary) [inline client/src/pages/CreateUserProfile.js]: Is there a reason for why getting the Auth0 user IDs and linking the Auth0 user IDs are separate API endpoints? Could we not combine them into one call?

**Review by mannykary — CHANGES_REQUESTED** [review]
Overall LGTM, just a few minor comments.

File: backend/lib/components/Auth0.js
Comment (ddveloper) [reply backend/lib/components/Auth0.js]: Thanks Manny, it was for updating user's meta data, which was removed then, I will remove this also

File: backend/lib/endpoints/auth.js
Comment (ddveloper) [reply backend/lib/endpoints/auth.js]: will do 👍

File: client/src/pages/CreateUserProfile.js
Comment (ddveloper) [reply client/src/pages/CreateUserProfile.js]: Yes Manny, we were thinking to let user decide link or not at the beginning then decided not to do that  -> [link](https://fightpandemics.slack.com/archives/C014E5X0FNY/p1605111386057900?thread_ts=1604939904.055700&cid=C014E5X0FNY) <- these 2 endpoints can be combined together, I will combine them as one.

**Review by Zaydme — DISMISSED** [review]
LGTM! 🙌

**Review by mannykary — DISMISSED** [review]
lgtm

**Review by idjevm — DISMISSED** [review]
🔥 LGTM!

**Comment by idjevm** [comment]
@Zaydme et al, let's wait for some folks from the QA to test it out before we merge into staging...

**Review by rabbi08 — CHANGES_REQUESTED** [review]
@ddveloper 
1. The issue still exist as when user try to login using existing email, no message is displayed. Please see below GIF.
ezgif com-video-to-gif

2. Other than that I have seen the error message only once when I start the testing on this URL after that it is not working.
Now I am navigating to create profile and no error message -> when I try to login using Linkedin and Google, both.

3. Also, great if we can update the message displayed when user try to login using an email already registered.

"Your account is already linked with this email, please try again using other login mode."

**Comment by ddveloper** [comment]
Yes @rabbi08 once your accounts are linked together, you'll not be able to link them again, next time you sign up with social account, it will log you in directly, that's why you don't see the message again, I will need your steps to reproduce the error you saw

**Comment by rabbi08** [comment]
> Yes @rabbi08 once your accounts are linked together, you'll not be able to link them again, next time you sign up with social account, it will log you in directly, that's why you don't see the message again, I will need your steps to reproduce the error you saw

Agree, @ddveloper that's what the test is, but as I mentioned, looks like that now I am not able to login using my email. This account was registered as Google user and I tried it on linkedin for testing purpose and now no login work on my email.

Thanks

**Comment by rabbi08** [comment]
@ddveloper 
Below is the summary of the issues observed during the above mentioned flow:
1. After user clicks on ‘Ok’ on the user account linking pop up, the user should be logged in and should land on FP home page, like normal logins. 
Issue: Currently after clicking ‘Ok’ in the account linking pop up, user is navigated to Sign In screen again.

2. When user try to login using Email/pwd login mode and enter random pwd : valid error message should be displayed ONLY if we matching email in the DB.
i.e. if user have already created profile with Google/linkedin/Facebook -> Then try to login using email/pwd -> enter random password
Then system should know that this is NOT new user and they do NOT have pwd so the error message should be to guide them to ‘Join Now’. 
‘Login failed, reason: Create password using ‘Join Now’

3. ‘Edit Security’ option is not displayed for the user anymore.

Below are the test scenarios executed in the Sequence mentioned with '->':

1. Linkedin -> Enter ‘email/pwd’ -> create profile page -> enter required details -> click ‘Create profile’  -> profile created successfully
2. Now sign out -> click Sign In -> Google -> Create profile page with pop up ‘your accounts with same email were linked together, please log in with any of them’
3. Click ‘OK’ in the pop up 
4. User lands on ‘Sign In’ Page -> FAIL, user should be logged in with the same email
5. So, user again navigate to FP home page -> click on ‘Sign In’ -> google -> and they are signed in now as Google user

6. Now sign Out
7. Sign In -> Email address ‘same as used above’ -> enter RANDOM pwd
8. Error displayed:
                   Login failed, reason: Wrong email or password.

9. So now click on Join Now -> Enter same Email address used above -> enter Pwd and Confirm pwd
10. Click on Join Now -> email verification email is send 
11. Verified the email 
12. Click ‘Sign In’ Enter email/Pwd -> Click on ‘Sign In’ -> Create Profile page with pop up ‘your accounts with same email were linked together, please log in with any of them’
13. Click ‘OK’ in the pop up 
14. User lands on ‘Sign In’ Page -> FAIL, user should be logged in with the same email
15. So, user again navigate to FP home page -> click on ‘Sign In’ -> Enter email/Pwd -> Click on ‘Sign In’ 
    Now user is logged in.

**Comment by ddveloper** [comment]
Thanks @rabbi08 , about the 3 issues:
1. there was [discussion about this ](https://fightpandemics.slack.com/archives/C014E5X0FNY/p1604939904055700) and currently no simple solution yet

2. and 3 I made some changes, and let me know if they work as expected

**Comment by ddveloper** [comment]
@rabbi08 as we discussed yesterday, code was changed to `remind user for no account case` and the `Edit Security` changes were reverted since it will not work for linked account when `email/pwd` is not the primary account.
for the Facebook login in incognito mode, I saw staging build has no problem and other review build has same issue, could we review this after merging into staging, if the issue is still there then create an issue for that?

**Comment by rabbi08** [comment]
@ddveloper 
Verified this fix on the feature branch and accounts are getting linked if user is using same email address but different login modes: PASS

Below is the summary of the test scenarios executed:

- User sign Up through Social media channel:
1. Sign Up using Linkedin/google/Facebook (say 'Linkedin') : PASS
2. Try to Sign In using Google with same email address, used to signup in step#1
     After entering the required details -> user is redirected to 'create profile' page and following message is displayed.
    "Your account has been linked with your existing same email account. Please login again with any of them."
    Click 'Ok' -> user is directed to 'Sign In' Page : PASS

image

3. Now, user can login to FP application using 'same email address' and through different login modes.
4. User can try to login using direct email address and password even when the account have been created before for the same email through social media login mode. 
5. Following error message is displayed and user is guided to signup using 'Join Now' and the account get linked with existing account of that email address : PASS

Screen Shot 2020-12-13 at 11 58 18 PM


Note: So, all the accounts with same email address will be linked together with one account on FP: PASS

6. Following error message is displayed for new user, no account exist in FP: PASS

Screen Shot 2020-12-13 at 11 32 05 PM

**Review by rabbi08 — APPROVED** [review]

**Review by Zaydme — APPROVED** [review]
LGTM!

**Review by idjevm — APPROVED** [review]
LGMT 🔥 Amazing job @ddveloper !!!

**Review by sankethbe — APPROVED** [review]
I confirm that all the signup/login features mentioned above are working as intended. The user's existing FP account is also successfully linked when using same email ID for social login. Congrats @ddveloper @rabbi08 and team for finally solving this long pending issue!
Note: I've extensively tested using Google sign-in but not FB, LI

**Comment by rabbi08** [comment]
Note: Need to observe/restest the login scenario for Facebook after it is merge in STG. Will log a bug if the problem exist in STG as well.
LGTM. Thanks @ddveloper

**Pull request merged**
